### PR TITLE
KV specific NavigateInput component

### PR DESCRIPTION
**Pull request by Monkeychip**
This PR makes a KV specific `NavigateInput` component called `KvListFilter`. While the code and functionality are very similar to the old component, this is KV only and does not seem to suffer from the same focus issues as the current component. 

My goal here was to get us moving forward on this engine. While, I would like to revisit this component and make it so we don't query on every input change, we can address that later. I time-boxed myself, got this to a functionally working state and added test coverage.

This component's UX is smoother, but I can't promise there are not edge cases the old component covered that this one does. Some of the old code I didn't carry over, largely the `blur` event and the `shared prefix` calculations.

https://github.com/hashicorp/vault/assets/6618863/7bcea980-8302-4af6-aedc-c3ccbbe5a524

Tested as a binary to make sure no focus issues. ✅

### CODE REVIEW

File: ui/app/components/console/command-input.js
Comment (Monkeychip) [inline ui/app/components/console/command-input.js:1]: I needed access to keycodes.js so I moved it to the addon folder. For consistency, I changed `keycodes` to `key-codes` , which followed the other naming patterns.

File: ui/app/templates/vault/cluster/access/leases/list.hbs
Comment (Monkeychip) [inline ui/app/templates/vault/cluster/access/leases/list.hbs:1]: I hit a similar spacing issue in KV, which I fixed with a helper. I figured I'd clean this up while I was at it. Confirmed in the UI the spacing is still correct.

File: ui/lib/core/addon/components/navigate-input.js
Comment (Monkeychip) [inline ui/lib/core/addon/components/navigate-input.js:1]: Woot! finally got to a very old TODO

File: ui/lib/kv/addon/components/kv-list-filter.js [hunk 1]
```diff
@@ -0,0 +1,154 @@
...
+      this.router.transitionTo(this.kvRoute('list'), {
+        queryParams: { pageFilter: input },
+      });
+    }
+  }
+  /*
+  -handleKeyDown handles: tab, enter, backspace and escape. Ignores everything else.
+**/
+  @action
+  handleKeyDown(event) {
```
Comment (hashishaw) [inline ui/lib/kv/addon/components/kv-list-filter.js:77]: This method feels pretty overloaded with logic. I wonder if we can split out the specific key handling to other methods (`handleTab`, `handleEnter`) etc and this method would only be in charge of checking if the key event is one of the special keys, and preventing the default behavior if it is

File: ui/lib/kv/addon/components/page/list.hbs [hunk 2]
```diff
@@ -62,13 +67,17 @@
...
-    data-test-secret-list-empty-state
-    @title="No secrets yet"
-    @message="When created, secrets will be listed here. Create a secret to get started."
-  >
-    <LinkTo class="has-top-margin-xs" @route="create">
-      Create secret
-    </LinkTo>
-  </EmptyState>
+  {{#if @filterValue}}
+    <EmptyState @title="There are no secrets matching &quot;{{@filterValue}}&quot;. Press enter to add one." />
```
Comment (hashishaw) [inline ui/lib/kv/addon/components/page/list.hbs:71]: "Please enter to add one" is sort of confusing language. Maybe ask them to clear the filter instead?

File: ui/lib/kv/addon/routes/list-directory.js [hunk 3]
```diff
@@ -14,25 +14,40 @@ export default class KvSecretsListRoute extends Route {
   @service router;
   @service secretMountPath;
 
-  getPathToSecretFromUrl() {
-    const { path_to_secret } = this.paramsFor('list-directory');
-    return path_to_secret ? normalizePath(path_to_secret) : '';
-  }
+  queryParams = {
+    // TODO kv engine cleanup: remove by filtering in the component
```
Comment (hashishaw) [inline ui/lib/kv/addon/routes/list-directory.js:18]: Not sure we can do this, especially if we are going to be using `lazyPaginatedQuery` (which we likely will need to do -- KV secret lists can be quite long IIRC)

File: ui/lib/kv/addon/components/kv-list-filter.js [hunk 4]
```diff
@@ -0,0 +1,154 @@
...
+import Component from '@glimmer/component';
+import { inject as service } from '@ember/service';
+import { action } from '@ember/object';
+import keys from 'core/utils/key-codes';
+import { keyIsFolder, parentKeyForKey, keyWithoutParentKey } from 'core/utils/key-utils';
+import escapeStringRegexp from 'escape-string-regexp';
+import { tracked } from '@glimmer/tracking';
+
+/**
+ * KvListFilter component is similar to the NavigateInput component but meets the specific routing needs for the KV Ember engine.
```
Comment (hashishaw) [inline ui/lib/kv/addon/components/kv-list-filter.js:15]: Maybe let's describe this outside of NavigateInput (since we eventually want to remove that)

**Review by hashishaw — COMMENTED** [review]
Thanks for this, and the tests! A few minor comments

File: ui/lib/kv/addon/components/kv-list-filter.js (hunk 1 above)
Comment (Monkeychip) [reply ui/lib/kv/addon/components/kv-list-filter.js:77]: Good idea. I'll amend.

File: ui/lib/kv/addon/components/page/list.hbs (hunk 2 above)
Comment (Monkeychip) [reply ui/lib/kv/addon/components/page/list.hbs:71]: I was trying to hint at hitting "enter" to add a new secret. But I think that's confusing so I'll just remove the sentence.

File: ui/lib/kv/addon/routes/list-directory.js (hunk 3 above)
Comment (Monkeychip) [reply ui/lib/kv/addon/routes/list-directory.js:18]: Yeah, I kind of wondered. I played around with this and couldn't get it to work without pagination so I can imagine it's only going to get harder. I'll remove the TODO. I've got some notes about addressing this in a nice-to-have jira ticket.

**Review by hashishaw — APPROVED** [review]
Thank you!

**Pull request merged**